cgaal-cli/src/args.rs

```diff
@@ -1,86 +1,243 @@
+    CliOptions, FormulaFormat, ModelFormat, SearchStrategyOption, SubcommandOption,
+};
+use clap::{App, AppSettings, Arg, ArgMatches, SubCommand};
+use git_version::git_version;
+
+const PKG_NAME: &str = env!("CARGO_PKG_NAME");
+const AUTHORS: &str = env!("CARGO_PKG_AUTHORS");
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+const GIT_VERSION: &str = git_version!(fallback = "unknown");
+
+/// Define and parse command line arguments
+pub fn parse_arguments() -> Result<CliOptions, String> {
+    let version_text = format!("{} ({})", VERSION, GIT_VERSION);
+    let string = AUTHORS.replace(':', "\n");
+    let mut app = App::new(PKG_NAME)
+        .version(version_text.as_str())
+        .author(string.as_str())
+        .arg(
+            Arg::with_name("log_filter")
+                .short("l")
+                .long("log-filter")
+                .env("RUST_LOG")
+                .default_value("warn")
+                .global(true)
+                .help("Comma separated list of filter directives"),
+        )
+        .subcommand(
+            SubCommand::with_name("check")
+                .about("Checks satisfiability of an ATL property on a CGS")
+                .add_positional_model_path_arg()
+                .add_model_format_arg()
+                .add_positional_formula_arg()
+                .add_formula_format_arg()
+                .add_search_strategy_arg()
+                .add_witness_strategy_arg()
+                .add_quiet_arg()
+                .add_threads_arg(),
+        )
+        .subcommand(
+            SubCommand::with_name("index")
+                .about("Prints indexes of LCGS declarations")
+                .add_positional_model_path_arg(),
+        )
+        .subcommand(
+            SubCommand::with_name("analyse")
+                .about("Analyses a EDG generated from an ATL query and a CGS")
+                .add_positional_model_path_arg()
+                .add_model_format_arg()
+                .add_positional_formula_arg()
+                .add_formula_format_arg()
+                .add_positional_output_arg(true),
+        )
+        .setting(AppSettings::SubcommandRequiredElseHelp);
+
+    if cfg!(feature = "graph-printer") {
+        app = app.subcommand(
+            SubCommand::with_name("graph")
+                .about(
+                    "Outputs a Graphviz DOT graph of the EDG generated from an ATL query and a CGS",
+                )
+                .add_positional_model_path_arg()
+                .add_model_format_arg()
+                .add_positional_formula_arg()
+                .add_formula_format_arg()
+                .add_positional_output_arg(false),
+        );
+    }
+
+    let arg_matches = app.get_matches();
+
+    setup_tracing(&arg_matches)?;
+
+    let mut options = CliOptions::default();
+    match arg_matches.subcommand() {
+        ("check", Some(args)) => {
+            options.subcommand = SubcommandOption::Check;
+            options.use_global = false;
```

Comment:
It is very elegant that way, but I worry it will bite us in the ass later down the line. E.g. the second suggestion in #219 would then have the user specify multiple methods, `--method=bfs,bfs,dfs`, which just seems unintuitive. In any case "method" is too general. Slicing, quotienting, expression rewriting, etc. are also methods that we may add at some point with flags to turn them on/off. So `--algorithm` or something would be better. But again, is it future-proof?

It does not remove all argument conflicts since global does not support `--no-prioritized-backpropagation` (and other local-algorithm-specific flags in the future. Vice versa of global-specific).

Reply:
The alternatives are having subcommands for `check` that are `local` or `global` but that's verbose for our default algorithm, so in that case a `--global` flag for subcommand `check` is probably best.

Let's keep it as suggested but have it in mind for future expansion of features.